implemented/0021-transfer-files-over-mqtt.md

```diff
@@ -81,15 +81,15 @@ As an example of existing implementation we can look at AWS IoT Core [which prov
   * segment, if it's a segment transfer message
   * whole file, if it's a `fin` message
 * If checksum verification fails, Broker MUST reject the corresponding data
-* Client MUST use Topic starting with `$file/` to transfer files
-* Broker MUST NOT let clients subscribe to Topics starting with `$file/` topics
+* Client MUST use Topic starting with `$file/`(`$file-async/`) to transfer files.
+* Broker MUST NOT let clients subscribe to Topics starting with `$file/`(`$file-async/`) topics
 * Segment length can be calculated on the server side by subtracting the length of the [Variable Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901025) from the [Remaining Length](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901105) field that is in the [Fixed Header](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901021)
 
 ### Protocol flow
 
 Data is transferred in PUBLISH packets in the following order:
 
-1. `$file/{fileId}/init`
+1. `$file[-async]/{fileId}/init`
```

Comment:
Q: A pity that we'd need another system topic prefix for that. What do you think about moving those kind of things at the end of the topic, in a http-query-string fashion? I.e. `$/file/{fileId}/init/?mode=async`, or `$/file/{fileId}/init//mode=async` (somewhat easier to deconstruct), or something similar. This would also provide a kind of forward compatibility. Apparent downside is that commands with optional checksums at the end would be slightly more complex to parse / handle.

Reply:
I considered some variants with mode/etc encoding as key-value pairs. However, it finally seemed to me that encoding something complex into topics somewhat contradicts the MQTT idea of topics, which are built from separators and opaque tokens.
With `/mode=.../` we e.g. cannot match against mode with standard topic matchers (`+`, `#`), cannot impose ACL, etc.

Reply:
Fair point. What do you think about a kind of middle ground where we define _mode_ as additional "positional argument", so to say? This can still provide us with forward compat, admittedly somewhat limited. E.g. `$file/abc-def/fin/4242//async` for a FIN command, where checksum is not specified. We could assume sync mode by default and say that `$file/abc-def/fin/4242//` is equivalent to `$file/abc-def/fin/4242` as a backward compat measure.